File: frontend/src/api/task.ts
import client from './client'

export interface TaskClinicalInfoSnapshot {
  [key: string]: any
}

export interface TaskItem {
  id: string
  task_type: string
  status: string
  project_type: string | null
  project_name?: string | null
  sample_id?: string | null
  sample_label?: string | null
  upload_id?: string | null
  original_filename?: string | null
  clinical_info_snapshot?: TaskClinicalInfoSnapshot
  total_files: number
  completed_files: number
  failed_files: number
  output_path?: string | null
  created_at: string | null
  started_at: string | null
  completed_at: string | null
  duration_seconds: number | null
  errors: string[]
  warnings?: string[]
  editable?: boolean
}

export interface TaskListResponse {
  items: TaskItem[]
  total: number
  page: number
  page_size: number
}

export interface TaskStats {
  total: number
  completed: number
  failed: number
  running: number
  pending: number
}

export interface TaskResultItem {
  id: number
  task_id: string
  file_index: number
  excel_filename: string
  status: string
  output_path?: string | null
  duration_seconds?: number | null
  errors: string[]
  warnings: string[]
  validation_summary?: Record<string, any>
}

export const taskApi = {
  async list(params: { status?: string; task_type?: string; page?: number; page_size?: number } = {}): Promise<TaskListResponse> {
    const { data } = await client.get('/tasks', { params })
    return data.data
  },

  async getStats(): Promise<TaskStats> {
    const { data } = await client.get('/tasks/stats')
    return data.data
  },

  async get(taskId: string): Promise<TaskItem> {
    const { data } = await client.get(`/tasks/${taskId}`)
    return data.data
  },

  async getResults(taskId: string): Promise<TaskResultItem[]> {
    const { data } = await client.get(`/tasks/${taskId}/results`)
    return data.data
  },

  async cancel(taskId: string): Promise<void> {
    await client.delete(`/tasks/${taskId}`)
  },

  async remove(taskId: string): Promise<void> {
    await client.delete(`/tasks/${taskId}/record`)
  },
}

File: backend/app/models/task.py
"""Task and TaskResult ORM models."""

from datetime import datetime

from sqlalchemy import DateTime, Float, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class Task(Base):
    __tablename__ = "tasks"

    id: Mapped[str] = mapped_column(String(36), primary_key=True)  # UUID
    user_id: Mapped[int | None] = mapped_column(Integer, nullable=True)
    upload_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    task_type: Mapped[str] = mapped_column(String(20), nullable=False)  # single | batch
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending")
    project_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    template_path: Mapped[str | None] = mapped_column(String(500), nullable=True)
    clinical_info_snapshot: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON

    # Results
    output_path: Mapped[str | None] = mapped_column(String(500), nullable=True)
    preview_path: Mapped[str | None] = mapped_column(String(500), nullable=True)
    context_json_path: Mapped[str | None] = mapped_column(String(500), nullable=True)

    # Batch-specific
    total_files: Mapped[int] = mapped_column(Integer, default=1)
    completed_files: Mapped[int] = mapped_column(Integer, default=0)
    failed_files: Mapped[int] = mapped_column(Integer, default=0)

    # Timing
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    started_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    duration_seconds: Mapped[float | None] = mapped_column(Float, nullable=True)

    # Error info
    errors: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    warnings: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON


class TaskResult(Base):
    __tablename__ = "task_results"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    task_id: Mapped[str] = mapped_column(String(36), nullable=False)
    file_index: Mapped[int] = mapped_column(Integer, nullable=False)
    excel_filename: Mapped[str] = mapped_column(String(255), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False)
    output_path: Mapped[str | None] = mapped_column(String(500), nullable=True)
    duration_seconds: Mapped[float | None] = mapped_column(Float, nullable=True)
    errors: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    warnings: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    validation_summary: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
